docs/technology.md

```diff
@@ -43,3 +43,119 @@ The slowest step in the pipeline is the graph database writer.
+good candidate for a first pass at migrations.
+
+- **Adding a property with a non-trivial default value** Adding a property with a default value that is not a constant. This is not supported because it would require evaluating python code per item being migrated which is not supported due to the additional complexity it would introduce. In frameworks like Django, this is supported because the framework can rely on the ORM to provide a consistent interface for evaluating the default value. In nodestream, there is no such interface.
+
+### Migration Anatomy 
+
+A migration is a yaml file that contains a list of operations to be performed on the database. 
+Each operation represents a change to the database schema (e.g creating and index, creating a node type, etc).
+The operations are executed in the order they are defined in the file. 
+The file is named with a timestamp and a description of the migration.
+While the timestamp will usually correspond to the order the migrations are applied, a list of `dependencies` is used to explicitly define the order of execution.
+
+Internally, `nodestream` will build a migration graph from all migrations in the project and their dependencies.
+It will then execute the migrations in topological order to ensure that all dependencies are met before a migration is applied.
+
+### Change Detection
+
+To determine if a migration needs to be created, `nodestream` will compare the current state of the project to the state 
+described in the migrations. 
+If there are any differences, a new migration will be created with the appropriate operations to bring the migrations up to date with the current state of the project.
+The design of this process is inspired by [Django Migrations](https://docs.djangoproject.com/en/5.0/topics/migrations/). 
+Specifcically, the [Autodetector](https://github.com/django/django/blob/main/django/db/migrations/autodetector.py) class. 
+
+
+### Developer Experience
+
+The developer experience has several core goals:
+
+- **Low friction.** The developer should be able to make changes to the database schema without having to manually manage the migrations.
+- **Fast.** The developer should be able to make changes to the database schema without having to wait for the migrations to be applied.
+- **Safe.** The developer should be able to make changes to the database schema without having to worry about breaking the pipeline.
+
+To achieve these goals, `nodestream` provides a set of commands that allow the developer to manage the migrations.
+
+- `nodestream migratations make [--pipeline <pipeline_name> --scope <scope_name>]` - Diff the current state of the project with the migrations and create a new migration if there are any differences. If `--pipeline` is specified, only migrations for that pipeline will be created. If `--scope` is specified, only migrations for that scope will be created.
```

Comment:
add a `--dry-run` option to see what the migration would consist of if required.